orttraining/orttraining/training_ops/cuda/optimizer/adamw/adamw.cc

```diff
@@ -0,0 +1,148 @@
+                                                             source_tensor.Shape(), GetAllocatorPtr(ctx));
+      CUDA_RETURN_IF_ERROR(cudaMemcpyAsync(target_tensor->MutableDataRaw(),
+                                           source_tensor.DataRaw(),
+                                           source_tensor.SizeInBytes(),
+                                           cudaMemcpyDeviceToDevice, steam));
+      updated_values->Add(std::move(*target_tensor));  // Add will check for type consistency
+    }
+  }
+
+  return Status::OK();
+}
+
+Status AdamWOptimizer::ComputeInternal(OpKernelContext* ctx) const {
+  const Tensor* learning_rate = ctx->Input<Tensor>(0);
+  const Tensor* step = ctx->Input<Tensor>(1);
+  const TensorSeq* weights = ctx->Input<TensorSeq>(2);
+  const TensorSeq* gradients = ctx->Input<TensorSeq>(3);
+  const TensorSeq* momentums_1 = ctx->Input<TensorSeq>(4);
+  const TensorSeq* momentums_2 = ctx->Input<TensorSeq>(5);
+
+  size_t num_of_weights = weights->Size();
+  size_t num_of_gradients = gradients->Size();
+  size_t num_of_momentums_1 = momentums_1->Size();
+  size_t num_of_momentums_2 = momentums_2->Size();
+
+  // Check the number of weights, gradients, momentums matchs.
+  ORT_ENFORCE(num_of_weights > 0, "invalid count of tensors in Seq<Tensor>.");
+  ORT_ENFORCE(num_of_weights == num_of_gradients, "number of weights and gradients mismatch.");
+  ORT_ENFORCE(num_of_gradients == num_of_momentums_1, "number of gradients and momentums_1 mismatch.");
+  ORT_ENFORCE(num_of_momentums_1 == num_of_momentums_2, "number of momentums_1 and momentums_2 mismatch.");
+
+  std::vector<int> tensor_sizes(num_of_weights);
+  std::vector<std::vector<void*>> grouped_tensor_pointers(num_of_weights);
+
+  for (size_t i = 0; i < num_of_weights; ++i) {
+    const Tensor& weight_tensor = weights->Get(i);
+    const Tensor& gradient_tensor = gradients->Get(i);
+    const Tensor& momentum_1_tensor = momentums_1->Get(i);
+    const Tensor& momentum_2_tensor = momentums_2->Get(i);
+
+    // Check the weight/gradient/momentums at the same index should have same shape.
+    ORT_ENFORCE(weight_tensor.Shape() == gradient_tensor.Shape(), "shape of weight and gradient mismatch.");
+    ORT_ENFORCE(gradient_tensor.Shape() == momentum_1_tensor.Shape(), "shape of gradient and momentum_1 mismatch.");
+    ORT_ENFORCE(momentum_1_tensor.Shape() == momentum_2_tensor.Shape(), "shape of momentum_1 and momentum_2 mismatch.");
+
+    // Currently we only support float data types.
+    ORT_ENFORCE(weight_tensor.IsDataType<float>() &&
+                    gradient_tensor.IsDataType<float>() &&
+                    momentum_1_tensor.IsDataType<float>() &&
+                    momentum_2_tensor.IsDataType<float>(),
+                "Only float data type support for Adam cuda kernel.");
+
+    tensor_sizes[i] = static_cast<int>(weight_tensor.Shape().Size());
+
+    grouped_tensor_pointers[i] = {
```

Comment:
I don't know if it is feasible, but what if we order the inputs this way itself? So the adamw op will have variadic number of inputs, and each input is a sequence of weight, gradient, moment1, moment2. This might save this sequential reordering cost of O(n)

Reply:
A valid SeqTensor should contain data in same data types, so it is feasible when we want to use different data types for weights and gradients, or others.

Reply:
The sequential reordering cost should be fine I think,  we can re-visit it when it is really a big cost during our profiling.